sopel/bot.py

```diff
@@ -898,6 +898,15 @@ def dispatch(self, pretrigger):
         user_obj = self.users.get(nick)
         account = user_obj.account if user_obj else None
 
+        if "time" in pretrigger.tags and pretrigger.sender in self.channels:
+            time = datetime.strptime(
+                pretrigger.tags["time"],
+                "%Y-%m-%dT%H:%M:%S.%f%z",
```

Comment:
Is there some reason to parse `%z` instead of a literal `Z`? The wire format prescribed by [the `server-time` spec](https://ircv3.net/specs/extensions/server-time) specifically says UTC, with a `Z` suffix, no other timezone. Tested against `tilde.chat`, got tags like the following in Sopel's raw log:

`@time=2021-06-09T02:30:54.891Z`

The use of `%z` is wonky on Python 2.7, at least, as I found out when rewriting the caching logic for `tld`, and for obvious reasons this will have to wait until Sopel 8.0 (instead of 7.1.1) if it doesn't work properly on py2. (Though I had this thought before remarking that a new field is added to the `Channel` type… Hmm. Maybe this should just be an 8.0 feature.)

Reply:
Yup, go with literal Z, the spec is quite clear on that:

> The value of the tag MUST have the following syntax:
> 
>     <value> ::= YYYY-MM-DDThh:mm:ss.sssZ

Which is, indeed, ISO 8601:2004(E) 4.3.2, a commonly used format for the Zulu timezone (UTC+0). There is no reason to use `%z` in that context.

Reply:
I forgot 7.1 still supports py2.7 and we're not free of it yet -.-
Using timezone-naïve objects is incorrect, and `datetime` sucks so the other ways to make the object timezone-aware (even UTC) are dumb. If this PR is going for 7.1 I'll change it, otherwise `%z` is best.

Reply:
> If this PR is going for 7.1 I'll change it, otherwise `%z` is best.

I disagree: go with literal `Z` in either case. The spec already tell you which timezone it should be (UTC+0). Forcing the timestamp to be in UTC+0 is a good thing, anything that isn't according to 

Also, what is dumb with the snippet of code?

```python
pytz.utc.localize(
    datetime.strptime(
        pretrigger.tags["time"], '%Y-%m-%dT%H:%M:%S.%fZ'))
```

I know you are annoyed with datetime, but right now, you are being stubborn about the wrong thing.

Reply:
Was going to say, you can get an aware `datetime` object without having to use `%z`. The `tld` plugin does it in a slightly different (kludgier) way, but still:

https://github.com/sopel-irc/sopel/blob/e2d8439caae33ed3dfe08f0c164d1e8323fd6f7d/sopel/modules/tld.py#L33-L35

Anyway, I thought about it some more and this should go in 8.0. It's a significant behavior change, not appropriate for 7.1.1. Holding it also absolves @half-duplex of any need to support Python 2.7, and removes the risk of anyone getting mad at the project (meaning me) for adding a "new feature" (the added field on `Channel` objects) in a maintenance release.

Reply:
> there should be a tzinfo= kwarg to strptime and many others

I think this every time I use `strptime()`. It's silly that there isn't one.

Maybe Python 4 will make datetime objects default to aware in the system's local TZ. We can dream.